rest/method_item_put_test.go

```diff
@@ -294,6 +363,179 @@ func TestPutItem(t *testing.T) {
 			ResponseBody: `{"id": "1", "foo": "3"}`,
 			ExtraTest:    checkPayload("foo.sub", "1", map[string]interface{}{"id": "1", "foo": "3"}),
 		},
+
+		`pathID:not-found,body:valid,default:missing`: {
+			Init: initForDefaultField(now, yesterday),
+			NewRequest: func() (*http.Request, error) {
+				body := bytes.NewReader([]byte(`{"foo": "baz"}`))
+				return http.NewRequest("PUT", "/foo/66", body)
+			},
+			ResponseCode: http.StatusCreated,
+			ResponseBody: `{"id": "66", "foo": "baz", "bar": "default"}`,
+			ExtraTest:    checkPayload("foo", "66", map[string]interface{}{"id": "66", "foo": "baz", "bar": "default"}),
+		},
+		`pathID:not-found,body:valid,default:set`: {
+			Init: initForDefaultField(now, yesterday),
+			NewRequest: func() (*http.Request, error) {
+				body := bytes.NewReader([]byte(`{"foo": "baz", "bar": "value"}`))
+				return http.NewRequest("PUT", "/foo/66", body)
+			},
+			ResponseCode: http.StatusCreated,
+			ResponseBody: `{"id": "66", "foo": "baz", "bar": "value"}`,
+			ExtraTest:    checkPayload("foo", "66", map[string]interface{}{"id": "66", "foo": "baz", "bar": "value"}),
+		},
+		`pathID:found,body:valid,default:missing`: {
+			Init: initForDefaultField(now, yesterday),
+			NewRequest: func() (*http.Request, error) {
+				body := bytes.NewReader([]byte(`{"foo": "baz"}`))
+				return http.NewRequest("PUT", "/foo/1", body)
+			},
+			ResponseCode: http.StatusOK,
+			ResponseBody: `{"id": "1", "foo": "baz"}`,
```

Comment:
I believe this request should set `"bar"` to `"default"` (no matter what the original value is).

It should probably be possibly be possible to clear via the new JSON PATCH request syntax, but for PUT, I think the result should be the same  for the first an second request, if possible.

Reply:
This is where the document is `found`. So `Default` field might have been deleted in a previous call, I don't think we need to reset it to Default. I believe (Default) should be applicable only for new documents.

Reply:
I have the following concerns:
1. Just _clearing_ the field, makes it unsafe to repeat a PUT call (when we don't know if the original item has been created or not).
2. Keeping the _original_ value becomes very inconsistent with fields that don't have a default.

Consider the following use-case: A smart home application of some sort, starts up by reading a config file, and calling "PUT" to create several "lights" with values set in the config file. When a value is not set in the config file, we want to use the server defaults.

With the sematics as described in this test, the first time we start the application, it will work, but for the second boot, it will _clear_ the fields it set for the first start. Same for any "lights" or "devices" we add to the config file after that; they will only remain valid for the first boot, and then end up clearing default values.

PUT is meant to hold _replace_ semantics; we are not updating a resource at a given URL, we are simply replacing it. for selective update, there is PATCH. When doing so, setting default values makes sense in my mind. There then are some exceptions, such as read-only fields, e.g. "createdAt", that I think it makes sense to still persist.

Clearing some field with a default value, is an extreme special case, and even if you are allowed to do so via a future JSON PATCH request, a PUT request at the URL should still "replace" the original resource.

Reply:
This is how `Default` is described in the README:
```
Default | 
The value to be set when resource is created and the client didn’t provide a value for the field. The content of this variable must still pass validation.
```
So it is tailored for resource creation, and its subsequent update is not mentioned. My code doesn't touch `ReadOnly` code path I presume. Deleting Read-Only field will bail out on Validation.